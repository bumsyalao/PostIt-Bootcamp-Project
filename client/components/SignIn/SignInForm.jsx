/* global Materialize */
import React from 'react';
import PropTypes from 'prop-types';
import { withRouter, Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { userSignInRequest } from '../../actions/SignInAction';

/**
 * @class SignInForm
 * @extends {React.Component}
 */
class SignInForm extends React.Component {
  /**
   * Creates an instance of SignInForm.
   * Binds class methods.
   * @param {any} props
   *
   * @memberOf SignInForm
   */
  constructor(props) {
    super(props);
    this.state = {
      loggedIn: false,
      username: '',
      password: ''
    };
    this.onChange = this.onChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }
    /**
   *
   * Check if user is authenticated
   * Directs user to homepage if user is already signed in
   *
   * @memberOf Homepage
   */
  componentDidMount() {
    if (this.props.access.isAuthenticated) {
      this.props.history.push('/homepage/welcome-page');
    }
  }
  /**
   * Sets the event value to the state
   * @return {void}
   * @param {Object} event The event of the HTML component
   *
   * @memberOf SignInForm
   */
  onChange(event) {
    this.setState({ [event.target.id]: event.target.value });
  }
  /**
   * Validates the SignInForm
   * Makes an action call to sign in the user
   * Toasts the error/success message
   * @return {void}
   *
   * @memberOf SignInForm
   */
  onSubmit() {
    const SigninInfo = {
      username: this.state.username,
      password: this.state.password
    };
    this.props.userSignInRequest(SigninInfo)
      .then((res) => {
        if (this.props.access.message) {
          this.props.history.push('/homepage/welcome-page');
          Materialize.toast(res.message, 5000, 'green');
        } else {
          Materialize.toast(`Login Failed: ${res.response.data.message}`, 5000, 'red');
        }
      })
      .catch((error) => {
        Materialize.toast(error.response.data.message, 5000, 'red');
      });
  }

  /**
   *
   * Renders SignInForm component
   * @returns SignInForm HTML component
   *
   * @memberOf SignInForm
   */
  render() {
    return (
      <div className="form-margin">
      <div><h3 className="form-header">LOGIN</h3></div>
        <div className="col s12 container">
          <div className="input-field col s6 offset-s3">
            <i className="material-icons prefix">account_circle</i>
            <input
              id="username"
              value={this.state.username}
              onChange={this.onChange}
              name="username"
              type="text"
              className="validate"
              required
            />
            <label className="active" htmlFor="credential">
              username
            </label>
          </div>
          <div className="input-field col s6 offset-s3">
            <i className="material-icons prefix">lock</i>
            <input
              id="password"
              value={this.state.password}
              onChange={this.onChange}
              name="password"
              type="password"
              className="validate"
              required
            />
            <label className="active" htmlFor="password">
              password
            </label>
          </div>
          <button
            onClick={this.onSubmit}
            id="submit-signin"
            className=
            "btn waves-effect waves-light col s6 offset-s3 red lighten-2"
            name="action"
          >
            Login<i className="material-icons right">send</i>
          </button>
        </div>
        <div className="col s6 offset-s3 forgot-password">
          <Link to="/forgot-password">Forgot Password?</Link>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  access: state.access
});

SignInForm.propTypes = {
  userSignInRequest: PropTypes.func.isRequired
};

export default connect(mapStateToProps, { userSignInRequest })(
  withRouter(SignInForm)
);
