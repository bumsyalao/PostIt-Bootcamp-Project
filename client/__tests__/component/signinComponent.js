
/* global expect jest test */
import React from 'react';
import Adapter from 'enzyme-adapter-react-15';
import { shallow, configure } from 'enzyme';
import toJson from 'enzyme-to-json';
import SignInForm from '../../components/SignIn/SignInForm.jsx';

configure({ adapter: new Adapter() });

describe('SignInForm Component', () => {
  const spy = jest.fn();
  beforeEach(() => {
    global.Materialize = { toast: jest.fn() };
  });
  afterEach(() => {
    spy.mockReset();
  });
  const props = {
    match: {
      path: ''
    },
    access: {
      isAuthenticated: false
    },
    history: {
      push: spy
    },
    signInUser: jest.fn(() => {
      return Promise.resolve();
    })
  };
  const component = shallow(
    <SignInForm {...props} />);
  test('should match the SignInForm snapshot', () => {
    const tree = toJson(component);
    expect(component.find('input').length).toEqual(2);
    expect(tree).toMatchSnapshot();
  });
  test('it should set state when onChange function is called', () => {
    component.instance().onChange(
      { target: { value: 'banku', name: 'username' } });
    component.instance().onChange(
      { target: { value: 'banku123', name: 'password' } });
    expect(component.state('username')).toEqual('banku@gmail.com');
    expect(component.state('password')).toEqual('banku123');
  });
  test(
    'it should submit fields in state when onSubmit function is called', () => {
      jest.spyOn(component.instance(), 'onSubmit');
      component.instance().onChange(
        { target: { value: 'banku', name: 'username' } });
      component.instance().onChange(
        { target: { value: 'banku123', name: 'password' } });
      component.find('#submit-signin').simulate('click');
      expect(component.find('#submit-signin').length).toEqual(1);
      expect(component.instance().onSubmit.mock.calls.length).toEqual(1);
    });
  test('onSubmit function should run', () => {
    component.instance().onChange(
      { target: { value: 'banku', name: 'username' } });
    component.instance().onChange(
      { target: { value: 'banku123', name: 'password' } });
    expect(component.state('username')).toEqual('banku');
    expect(component.state('password')).toEqual('banku123');
    const newspy = jest.spyOn(component.instance(), 'onSubmit');
    component.instance().onSubmit();
    expect(newspy).toHaveBeenCalled();
  });
});
