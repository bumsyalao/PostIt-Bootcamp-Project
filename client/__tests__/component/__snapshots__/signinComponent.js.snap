// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignInForm Component should match the SignInForm snapshot 1`] = `
<div
  className="form-margin"
>
  <div
    className="col s12 container"
  >
    <div
      className="input-field col s6 offset-s3"
    >
      <i
        className="material-icons prefix"
      >
        account_circle
      </i>
      <input
        className="validate"
        id="username"
        name="username"
        onChange={[Function]}
        required={true}
        type="text"
      />
      <label
        className="active"
        htmlFor="credential"
      >
        username
      </label>
    </div>
    <div
      className="input-field col s6 offset-s3"
    >
      <i
        className="material-icons prefix"
      >
        lock
      </i>
      <input
        className="validate"
        id="password"
        name="password"
        onChange={[Function]}
        required={true}
        type="password"
      />
      <label
        className="active"
        htmlFor="password"
      >
        password
      </label>
    </div>
    <button
      className="btn waves-effect waves-light col s6 offset-s3 red lighten-2"
      id="submit-signin"
      name="action"
      onClick={[Function]}
    >
      Login
      <i
        className="material-icons right"
      >
        send
      </i>
    </button>
  </div>
  <div
    className="col s6 offset-s3 forgot-password"
  >
    <Link
      replace={false}
      to="/forgot-password"
    >
      Forgot Password?
    </Link>
  </div>
</div>
`;
